src/math/exp_fcn.c

```diff
@@ -0,0 +1,171 @@
+ * Multiplication of two signed int64 numbers
+ * Return Type	: void
+ */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
+{
+	uint64_t absin0;
+	uint64_t absin1;
+	uint64_t in0hi;
+	uint64_t in0lo;
+	uint64_t in1hi;
+	uint64_t prodhilo;
+	uint64_t prodlohi;
+	// bounds
+	absin0 = ABS(in_0);
+	absin1 = ABS(in_1);
+	// High and low bytes
+	in0hi = absin0 >> 32;
+	in0lo = absin0 & UINT32_MAX;
+	in1hi = absin1 >> 32;
+	absin0 = absin1 & UINT32_MAX;
+	// multiplication
+	prodhilo = in0hi * absin0;
+	prodlohi = in0lo * in1hi;
+	absin0 *= in0lo;
+	absin1 = 0;
+	in0lo = absin0 + (prodlohi << 32);
+	if (in0lo < absin0)
+		absin1 = 1;
+
+	absin0 = in0lo;
+	//shift and add
+	in0lo += prodhilo << 32;
+	if (in0lo < absin0)
+		absin1++;
+
+	absin0 = absin1 + in0hi * in1hi + (prodlohi >> 32) + (prodhilo >> 32);
+	if (in_0 != 0 && in_1 != 0 && (in_0 > 0) != (in_1 > 0)) {
+		absin0 = ~absin0;
+		in0lo = ~in0lo;
+		in0lo++;
+		// counter
+		if (in0lo == 0)
+			absin0++;
+	}
+	// pointer- output high and low bytes
+	*ptroutbitshi = absin0;
+	*ptroutbitslo = in0lo;
+}
+
+/* f(x) = a^x, x is variable and a is base
+ *
+ * Arguments	: int32_t x(Q4.28)
+ * input range -5 to 5
+ *
+ * Return Type	: int32_t ts(Q9.23)
+ * output range 0.0067465305 to 148.4131488800
+ *+------------------+-----------------+--------+--------+
+ *| x		     | ts (returntype) |   x    |  ts    |
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *|WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *| 32 | 28  |  1    | 32 | 23 |   0   | 4.28   | 9.23   |
+ *+------------------+-----------------+--------+--------+
+ */
+int32_t exp_int32(int32_t x)
+{
+	/* LUT = ceil(1/factorial(n) * 2 ^ 63) */
+	static const int64_t iv[19] = {
+			4611686018427387904LL,
+			1537228672809129301LL,
+			384307168202282325LL,
+			76861433640456465LL,
+			12810238940076077LL,
+			1830034134296582LL,
+			228754266787072LL,
+			25417140754119LL,
+			2541714075411LL,
+			231064915946LL,
+			19255409662LL,
+			1481185358LL,
+			105798954LL,
+			7053264LL,
+			440829LL,
+			25931LL,
+			1441LL,
+			76LL,
+			4LL
+	};
+
+	int64_t dividend;
+	int32_t ts;
+	int32_t n;
+	int64_t qt;
+	uint64_t ou0Hi;
+	uint64_t ou0Lo;
+
+	dividend = NUM_Q14P18;
+	ts = 0;/* Q23.9 */
+	n = 0; /* Q5.26 */
+	// pre-computation of 1st & 2nd terms
+	ts =  TERMS_Q23P9; /* Q23.9 */
+
+	dividend = (x + (1 << 13)) >> 14; //x in Q50.14
+
+	wide_mul_s64(dividend, BIT_MASK_Q62P2, &ou0Hi, &ou0Lo);
+	qt = (ou0Hi << 46ULL) | (ou0Lo >> 18ULL);/* Q6.26 */
+	ts += (int32_t)((qt >> 35) +
+			 (int64_t)((qt & QUOTIENT_SCALE) != 0 ? (int32_t)1 : (int32_t)0));
+	dividend = lomul_s64_SR_sat_near(dividend, (int64_t)x);
+	do {
+		int64_t minus_qt;
+
+		wide_mul_s64(dividend, iv[n], &ou0Hi, &ou0Lo);
+		qt = (int64_t)(ou0Hi << 45ULL) | (ou0Lo >> 19ULL);
+
+		// sum of the remaining terms
+		ts += (int32_t)((qt >> 35ULL) +
```

Comment:
Could the ? 1 : 0 could be avoided by summing qt the bit and result directly or via some further logic operations without conditions?

Reply:
I'll think about your suggestion more.  Thanks

Reply:
I made the following adjustment and saw that it resulted in a failure of the accuracy test. 

`QUOTIENT_SCALE = 17179869184LL;`
`((qt & QUOTIENT_SCALE) ? 1 : 0)` with ` ((qt & QUOTIENT_SCALE) >> 18)`

following are the  debug logs
```
working case
	dividend     | 		iv[n]	    |		  ou0Hi  |	     ou0Lo     |	   qt 	      |		    ts 	    |	           x      |
---------------------+----------------------+--------------------+---------------------+----------------------+---------------------+---------------------+
               64000 |     4000000000000000 |              fcf3a4|               fcf394|      3200000000000000|              4400000|             b0000000|
     ffffffffffe0c000|      1555555555555555|              fcf3a4|               fcf394|      acaaaaaaaaaaaaab|             f9d55555|             b0000000|
               9c4000|       555555555555555|              fcf3a4|               fcf394|      682aaaaaaaaaaaa4|              6daaaaa|             b0000000|
     fffffffffcf2c000|       111111111111111|              fcf3a4|               fcf394|      97d555555555555b|             f9d55555|             b0000000|
              f424000|        2d82d82d82d82d|              fcf3a4|               fcf394|      56ce38e38e38e294|              4af1c71|             b0000000|
     ffffffffb3b4c000|         6806806806806|              fcf3a4|               fcf394|      c1fefbefbefbf487|             fceefbef|             b0000000|
            17d784000|          d00d00d00d00|              fcf3a4|               fcf394|      26c0a28a28a26368|              1c71040|             b0000000|
     fffffff88ca6c000|          171de3a556c7|              fcf3a4|               fcf394|      ea78894177a5e343|             ff162168|             b0000000|
           2540be4000|           24fc9f6ef13|              fcf3a4|               fcf394|       ac3bb5f4428dd7b|               6e98d4|             b0000000|
     ffffff45bc48c000|            35cc8acfea|              fcf3a4|               fcf394|      fb1b65033e268db7|             ffd20574|             b0000000|
          3a352944000|             47bb63bfe|              fcf3a4|               fcf394|       209eb3ea607333a|               1342dc|             b0000000|
     ffffedcf631ac000|              5849184e|              fcf3a4|               fcf394|      ff37430f4b636e84|             fffa2b3e|             b0000000|
         5af3107a4000|               64e5d2a|              fcf3a4|               fcf394|        47b131637393a9|                32164|             b0000000|
     fffe3940ad9cc000|                6b9fd0|              fcf3a4|               fcf394|      ffe81a44c81cb553|                 24ad|             b0000000|
        8e1bc9bf04000|                 6b9fd|              fcf3a4|               fcf394|         777ca81770755|                113a6|             b0000000|
     ffd39750f44ec000|                  654b|              fcf3a4|               fcf394|      fffdcdb608a3549a|                 cd5d|             b0000000|
       de0b6b3a764000|                   5a1|              fcf3a4|               fcf394|          9c3bc8d28273|                 e0e4|             b0000000|
     fba9c6e7dbb0c000|                    4c|              fcf3a4|               fcf394|      ffffd6cce19aa70f|                 dbbe|             b0000000|
     15af1d78b58c4000|                     4|              fcf3a4|               fcf394|           ad78ebc5ac6|                 dd19|             b0000000|

   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|
---------------+---------------+---------------+---------------+---------------+
    -5.00000000|     0.00674736|     0.00673795|  -100.52082740|     0.09417999|

non working case
	dividend     | 		iv[n]	    |		  ou0Hi  |	     ou0Lo     |	   qt 	      |		    ts 	    |	           x      |
---------------------+----------------------+--------------------+---------------------+----------------------+---------------------+---------------------+
               64000 |     4000000000000000 |              4ff1ac|               4ff19c|      3200000000000000|              4400000|             b0000000|
     ffffffffffe0c000|      1555555555555555|              4ff1ac|               4ff19c|      acaaaaaaaaaaaaab|             f9d55555|             b0000000|
               9c4000|       555555555555555|              4ff1ac|               4ff19c|      682aaaaaaaaaaaa4|              6daaaaa|             b0000000|
     fffffffffcf2c000|       111111111111111|              4ff1ac|               4ff19c|      97d555555555555b|             f9d65554|             b0000000|
              f424000|        2d82d82d82d82d|              4ff1ac|               4ff19c|      56ce38e38e38e294|              4b01c70|             b0000000|
     ffffffffb3b4c000|         6806806806806|              4ff1ac|               4ff19c|      c1fefbefbefbf487|             fcf0fbed|             b0000000|
            17d784000|          d00d00d00d00|              4ff1ac|               4ff19c|      26c0a28a28a26368|              1c9103e|             b0000000|
     fffffff88ca6c000|          171de3a556c7|              4ff1ac|               4ff19c|      ea78894177a5e343|             ff182166|             b0000000|
           2540be4000|           24fc9f6ef13|              4ff1ac|               4ff19c|       ac3bb5f4428dd7b|               7198d1|             b0000000|
     ffffff45bc48c000|            35cc8acfea|              4ff1ac|               4ff19c|      fb1b65033e268db7|             ffd50571|             b0000000|
          3a352944000|             47bb63bfe|              4ff1ac|               4ff19c|       209eb3ea607333a|               1742d8|             b0000000|
     ffffedcf631ac000|              5849184e|              4ff1ac|               4ff19c|      ff37430f4b636e84|             ffff2b39|             b0000000|
         5af3107a4000|               64e5d2a|              4ff1ac|               4ff19c|        47b131637393a9|                8215f|             b0000000|
     fffe3940ad9cc000|                6b9fd0|              4ff1ac|               4ff19c|      ffe81a44c81cb553|                624a7|             b0000000|
        8e1bc9bf04000|                 6b9fd|              4ff1ac|               4ff19c|         777ca81770755|                713a0|             b0000000|
     ffd39750f44ec000|                  654b|              4ff1ac|               4ff19c|      fffdcdb608a3549a|                7cd56|             b0000000|
       de0b6b3a764000|                   5a1|              4ff1ac|               4ff19c|          9c3bc8d28273|                7e0dd|             b0000000|
     fba9c6e7dbb0c000|                    4c|              4ff1ac|               4ff19c|      ffffd6cce19aa70f|                8dbb6|             b0000000|
     15af1d78b58c4000|                     4|              4ff1ac|               4ff19c|           ad78ebc5ac6|                9dd10|             b0000000| 
 
   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|
---------------+---------------+---------------+---------------+---------------+
    -5.00000000|     0.07705879|     0.00673795|   -23.05831837|   703.20845133|

```